### PR TITLE
Handlers in sub-directories are now supported.

### PR DESCRIPTION
**Pull request by justinpastoncooper**
This fixes https://github.com/seek-oss/serverless-haskell/issues/39.

I have tested both handlers at the same level, which was supported before, and handlers in sub-directories. Would you mind adding a suitable case to the integration test?

### CODE REVIEW

**Comment by justinpastoncooper** [comment]
Actually I will try adding a case.

File: serverless-plugin/index.js [hunk 1]
```diff
@@ -74,12 +74,22 @@ class ServerlessPlugin {
...
+            directoryArgs = [];
+        }
+
         return spawnSync(
             'stack',
             [
                 ...args,
                 ...dockerArgs,
                 ...this.custom.stackBuildArgs,
+                ...directoryArgs
```
Comment (koterpillar) [inline serverless-plugin/index.js]: Keep the trailing commas, please.

File: serverless-plugin/index.js [hunk 2]
```diff
@@ -91,74 +101,90 @@ class ServerlessPlugin {
...
+            const matches = handlerPattern.exec(func.handler);
 
-        // Exclude Haskell artifacts from uploading
-        service.package.exclude = service.package.exclude || [];
-        service.package.exclude = [
-            ...service.package.exclude,
-            ...ADDITIONAL_EXCLUDE,
-        ];
+            if (!matches) {
+                throw new Exception(`handler ${func.handler} was not of the form ${handlerPattern}`);
```
Comment (koterpillar) [inline serverless-plugin/index.js]: Regular expressions aren't easy to read, please provide an example instead.

File: serverless-plugin/index.js [hunk 3]
```diff
@@ -91,74 +101,90 @@ class ServerlessPlugin {
...
-            ]
-        ).stdout)['dependencies'][PACKAGE_NAME]['version'];
-
-        if (haskellPackageVersion != javascriptPackageVersion) {
-            this.serverless.cli.log(`Package version mismatch: NPM: ${javascriptPackageVersion}, Stack: ${haskellPackageVersion}. Versions must be in sync to work correctly.`);
-            throw new Error("Package version mismatch.");
-        }
+        service.getAllFunctions().forEach(funcName => {
+            const func = service.getFunction(funcName);
+            const handlerPattern = /((.*\/))?([^\./]*)\.(.*)/;
```
Comment (koterpillar) [inline serverless-plugin/index.js]: No use for double brackets around the directory name.

File: serverless-plugin/index.js [hunk 4]
```diff
@@ -91,74 +101,90 @@ class ServerlessPlugin {
...
-            ...ADDITIONAL_EXCLUDE,
-        ];
+            if (!matches) {
+                throw new Exception(`handler ${func.handler} was not of the form ${handlerPattern}`);
+            }
 
-        // Each package will have its own wrapper; remember its options to add
-        // to the handler template
-        const handlerOptions = {};
+            const directory = matches[2];
```
Comment (koterpillar) [inline serverless-plugin/index.js]: Why not `const [directory, packageName, executableName] = matches`?

File: serverless-plugin/index.js [hunk 5]
```diff
@@ -186,57 +213,61 @@ class ServerlessPlugin {
...
+            // Create a shim to start the executable and copy it to all the
+            // destination directories
+            const handlerTemplate = fs.readFileSync(TEMPLATE).toString('utf8');
+
+            Object.keys(handlerOptions).forEach(packageName => {
+                let handler = handlerTemplate + handlerOptions[packageName].map(
+                    ([executableName, options]) => `exports['${executableName}'] = wrapper(${JSON.stringify(options)});`
+                ).join("\n") + "\n";
+
+                const handlerFileName = this.buildHandlerFileName(directory, packageName);
```
Comment (koterpillar) [inline serverless-plugin/index.js]: Why isn't this adding `this.servicePath` anymore?

Comment (justinpastoncooper) [reply serverless-plugin/index.js]: I've changed the definition of buildHandlerFileName to include `this.servicePath`.

**Comment by justinpastoncooper** [comment]
I've addressed your comments and updated the documentation. Is my change to the documentation sufficient?

**Comment by koterpillar** [comment]
I've added the light version of the integration test on master, so that the error should now surface on Travis.

**Comment by koterpillar** [comment]
Thanks for fixing that. I'll try to add the subdirectory handler to the integration test and merge it then.

**Comment by justinpastoncooper** [comment]
Okay.

**Comment by koterpillar** [comment]
There is a problem which I've inadvertently ran into: handlers in different directories with identical package names clash. This isn't likely to happen in practice, but it shouldn't silently do the wrong thing anyway. I think the right solution here is to put each executable into a corresponding directory (next to its handler).

Feel free to make that change - the integration test should be testing for that now.

**Comment by justinpastoncooper** [comment]
I don't fully understand. `buildHandlerFileName` returns a file name in the corresponding directory. Could you clarify?

**Comment by koterpillar** [comment]
I mean the (Haskell) executable, not the handler. All the executables will be in the root directory of the archive, and if they have the same name, will overwrite each other.

**Comment by justinpastoncooper** [comment]
I was confused and you’re absolutely right. Thanks for making the change.

**Pull request merged**